src/main/java/org/apache/sysml/hops/rewrite/HopRewriteUtils.java

```diff
@@ -241,7 +241,24 @@ public static void addChildReference( Hop parent, Hop child, int pos ){
 		parent.getInput().add( pos, child );
 		child.getParent().add( parent );
 	}
-	
+
+	/**
+	 * Replace an old Hop with a replacement Hop.
+	 * If the old Hop has no parents, then return the replacement.
+	 * Otherwise rewire each of the Hop's parents into the replacement and return the replacement.
+	 * @param old To be replaced
+	 * @param replacement The replacement
+	 * @return replacement
+	 */
+	public static Hop replaceHop(final Hop old, final Hop replacement) {
+		final ArrayList<Hop> rootParents = old.getParent();
+		if (rootParents.isEmpty())
```

Comment:
well, it's a nice to have comment but it's generally useful to use common primitives for all rewrites (which makes later refactoring much simpler). Maybe you can simply add this empty check to rewireAllParentChildReferences (which would also work for empty parents) in order to avoid the unnecessary ArrayList creation?